In [23]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Problem 3

In [16]:
def logistic_mini_sgd(X, y, batch_size, learning_rate, max_iter=1000):
    n_samples, n_features = X.shape
    w = np.zeros(n_features)  # initialize weights with value zero

    def sigmoid(z):
        return 1 / (1 + np.exp(-z))

    def cross_entropy_loss(y_true, y_pred):
        return -np.mean(y_true * np.log(y_pred) + (1 - y_true) * np.log(1 - y_pred))

    # training loop
    for i in range(max_iter):
        # shuffle the data for each iteration
        indices = np.arange(n_samples)
        np.random.shuffle(indices)

        for batch_start in range(0, n_samples, batch_size):
            batch_indices = indices[batch_start:batch_start + batch_size]
            X_batch = X[batch_indices]
            y_batch = y[batch_indices]

            # predict using sigmoid
            y_pred = sigmoid(np.dot(X_batch, w))

            # find gradients
            error = y_pred - y_batch
            grad = np.dot(X_batch.T, error) / batch_size  # Average gradient over batch

            # update weights
            w = w - learning_rate * grad

    return w

# Problem 4

a)

In [5]:

bc = load_breast_cancer()

b)

In [9]:
X = bc.data
y = bc.target
X_train, X_rem, y_train, y_rem = train_test_split(X, y, train_size=0.75) # split the data to get training dataset
x_val, X_test, y_val, y_test = train_test_split(X_rem, y_rem, test_size=0.4) # split the remaining to get validation and test datasets
# scale = StandardScaler()
# X_train = scale.fit_transform(X_train)
# x_val = scale.transform(x_val)
# X_test = scale.transform(X_test)

c)

In [15]:
# combine training and validation sets
y_train_val = np.concatenate([y_train, y_val])

# count
count_train_val = np.bincount(y_train_val)

print(f"Class distribution in training (+ validation) set: {count_train_val}")

Class distribution in training (+ validation) set: [194 317]


d)

In [29]:
weights = logistic_mini_sgd(X_train, y_train, batch_size=10, learning_rate=0.01)

print("Trained Weights:", weights)

weights = logistic_mini_sgd(X_train, y_train, batch_size=32, learning_rate=0.01)

print("Trained Weights:", weights)

weights = logistic_mini_sgd(X_train, y_train, batch_size=32, learning_rate=0.001)

print("Trained Weights:", weights)

weights = logistic_mini_sgd(X_train, y_train, batch_size=5, learning_rate=0.0001)

print("Trained Weights:", weights)

<ipython-input-16-0a169bbda1e1>:8: RuntimeWarning: overflow encountered in exp
  return 1 / (1 + np.exp(-z))


Trained Weights: [ 2.13286971e+01 -2.00860450e+01  7.95153773e+01  7.33196947e+00
 -1.20079876e-01 -1.45582088e+00 -2.24860375e+00 -8.38478440e-01
 -2.58668001e-01  1.99581772e-04  5.16317250e-01 -4.42954485e-01
 -3.06439290e+00 -1.66622522e+01 -3.46137649e-02 -4.26790888e-01
 -5.55277669e-01 -1.11695371e-01 -1.04783249e-01 -3.87077179e-02
  2.19830717e+01 -4.21094100e+01  3.25960194e+01 -1.69557094e+01
 -3.61789739e-01 -5.28348372e+00 -6.54583365e+00 -1.65310737e+00
 -1.13310464e+00 -4.04023498e-01]
Trained Weights: [ 8.88974197e+00 -7.33865423e+00  3.85244111e+01  6.30000445e+00
 -8.89129999e-03 -4.38387600e-01 -7.13951634e-01 -2.65855890e-01
 -3.63688950e-02  1.93296815e-02  1.68016194e-01 -2.63721259e-01
 -9.94750187e-01 -1.14638262e+01 -9.80267436e-03 -1.33738457e-01
 -1.73925720e-01 -3.40526929e-02 -2.97209460e-02 -1.17915448e-02
  9.28361919e+00 -1.48735041e+01  2.48372853e+01 -1.24930295e+01
 -7.75420517e-02 -1.62538191e+00 -2.05412783e+00 -5.09473967e-01
 -2.91327365e-01 -1.03

e)

In [30]:
def sigmoid(z):
        return 1 / (1 + np.exp(-z))

def predict(X, w):
    probs = sigmoid(np.dot(X, w))
    return np.where(probs >= 0.5, 1, 0)  # Threshold at 0.5

# test prediction
y_pred_test = predict(X_test, weights)

# performance metrics
accuracy = accuracy_score(y_test, y_pred_test)
precision = precision_score(y_test, y_pred_test)
recall = recall_score(y_test, y_pred_test)
f1 = f1_score(y_test, y_pred_test)

print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-score: {f1}")

Accuracy: 0.9482758620689655
Precision: 0.9302325581395349
Recall: 1.0
F1-score: 0.963855421686747


f)

The trained weights in (d) show that smaller batch sizes and learning rates generally lead to better convergence.

With an accuracy of 94.82%, the model is classifying most points correctly.
The precision is 0.93, indicating that that when the model predicts a positive class, it is correct about 93% of the time.
The recall is 1.0, meaning that the model has successfully identifies all actual positive instances.
With a 0.964 F1-score, the model has a good overall performance.

The recall is 1.0, but with a slightly lower precision, the model might have slightly overpredicted the positive class. **bold text**